kore/src/Kore/Step/Simplification/AndTerms.hs

```diff
@@ -831,15 +831,23 @@ functionAnd
     -> TermLike variable
     -> Maybe (Pattern variable)
 functionAnd first second
-  | isFunctionPattern first, isFunctionPattern second =
-    return Conditional
-        { term = first  -- different for Equals
+  | not (isFunctionPattern first) || not (isFunctionPattern second) = empty
+  | App_ symbol _ <- first
+  , Symbol.isDeclaredFunction symbol
+  , TermLike.isNonSimplifiable second
+  , isFunctionPattern first, isFunctionPattern second
+  =
+    return (Pattern.withCondition second condition)
+  | App_ symbol _ <- second
+  , Symbol.isDeclaredFunction symbol
+  , TermLike.isNonSimplifiable first
```

Comment:
why is there an asymmetry between this case and the above? Above you are checking that the second is non-simplifiable and both the first and the second are non-simplifiable, here you are checking that the first is non-simplifiable and (implicitly) that at least one of the two is functional.

Also, both check that the top symbol for the perhaps-simplifiable term is a function. This is also mentioned in the PR description, but why would one check only the top symbol and not the entire term?

I mean, I know that this is a heuristic, but I'd like to understand it.

Something else that's confusing: I didn't yet figure out why this works for the example you provide in the PR description. I would have thought that `Lbl'Unds'ModuleInstCellMap'Unds'`, the top symbol in the first term, is a function symbol (or should be one, i.e. we should fix it if it isn't), both terms are functional, and the second one is non-simplifiable (since it's a builtin). If so, then it should match the second guard in this function, so we should not get an error.

Reply:
Ok, I think I found the answer to the last question - variables are not non-simplifiable.